nuxeo-runtime/nuxeo-stream/src/main/java/org/nuxeo/lib/stream/computation/AbstractBatchComputation.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     bdelbosc
+ */
+package org.nuxeo.lib.stream.computation;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import net.jodah.failsafe.Failsafe;
+
+/**
+ * An abstract {@link Computation} that processes records by batch with a retry mechanism.<br/>
+ *
+ * @since 10.3
+ */
+public abstract class AbstractBatchComputation extends AbstractComputation {
+
+    private static final Log log = LogFactory.getLog(AbstractBatchComputation.class);
+
+    public static final String TIMER_BATCH = "batch";
+
+    protected final ComputationPolicy policy;
+
+    protected final List<Record> batchRecords;
+
+    protected String currentInputStream;
+
+    protected boolean newBatch = true;
+
+    protected final long thresholdMillis;
+
+    /**
+     * Constructor
+     *
+     * @param name the name of the computation
+     * @param nbInputStreams the number of input streams
+     * @param nbOutputStreams the number of output streams
+     * @param policy the policy to manage batch, retries and fallback
+     */
+    public AbstractBatchComputation(String name, int nbInputStreams, int nbOutputStreams, ComputationPolicy policy) {
+        super(name, nbInputStreams, nbOutputStreams);
+        this.policy = policy;
+        thresholdMillis = policy.getBatchThreshold().toMillis();
+        batchRecords = new ArrayList<>(policy.batchCapacity);
+    }
+
+    /**
+     * Called when:<br>
+     * - the batch capacity is reached<br/>
+     * - the time threshold is reached<br/>
+     * - the inputStreamName has changed<br/>
+     * If this method raises an exception the retry policy is applied.
+     *
+     * @param context used to send records to output streams, note that the checkpoint is managed automatically.
+     * @param inputStreamName the input streams where the records are comming from
+     * @param records the batch of records
+     */
+    public abstract void batchProcess(ComputationContext context, String inputStreamName, List<Record> records);
+
+    /**
+     * Called when the retry policy has failed.
+     */
+    public abstract void batchFailure(ComputationContext context, String inputStreamName, List<Record> records);
+
+    @Override
+    public void init(ComputationContext context) {
+        context.setTimer(TIMER_BATCH, System.currentTimeMillis() + thresholdMillis);
+    }
+
+    @Override
+    public void processTimer(ComputationContext context, String key, long timestamp) {
+        if (!TIMER_BATCH.equals(key) || batchRecords.isEmpty()) {
+            return;
```

Comment:
Don't we need to set timer back if `batchRecords` is empty ?

Reply:
of course !